finality-aleph/src/data_io.rs

```diff
@@ -359,82 +1001,228 @@ where
+                highest_finalized: finalized_block.clone(),
+            });
+            let mut curr_block = best_block.clone();
+
+            if curr_block.num == finalized_block.num {
+                // We don't have anything to propose, we go ahead with an empty propasal.
+                *self.data_to_propose.lock() = AlephData::HeadProposal(AlephProposal::<B>::new(
+                    Vec::new(),
+                    self.min_block_num,
+                ));
+                return;
+            }
+            if curr_block.num < finalized_block.num {
+                // Because of the client synchronization, in extremely rare cases this could happen.
+                warn!(target: "afa", "Error updating data. curr block {:?}, finalized {:?}", curr_block, finalized_block);
                 return;
             }
+            let mut branch: Vec<B::Hash> = Vec::new();
+            while curr_block.num > finalized_block.num {
+                if curr_block.num - finalized_block.num
+                    <= <NumberFor<B>>::saturated_from(MAX_DATA_BRANCH_LEN)
+                {
+                    branch.push(curr_block.hash);
+                }
+                curr_block = get_parent(self.client.clone(), &curr_block)
+                    .expect("block of num >= 1 must have a parent")
+            }
+            if curr_block.hash == finalized_block.hash {
+                let num_last = finalized_block.num + <NumberFor<B>>::saturated_from(branch.len());
+                // The hashes in `branch` are ordered from top to bottom -- need to reverse.
+                branch.reverse();
+                *self.data_to_propose.lock() =
+                    AlephData::HeadProposal(AlephProposal::<B>::new(branch, num_last));
+            } else {
+                // By backtracking from the best block we reached a block conflicting with best finalized.
+                // This is most likely a bug, or some extremely unlikely synchronization issue of the client.
+                warn!(target: "afa", "Error updating data. Conflicting blocks: {:?}, finalized {:?}", curr_block, finalized_block);
+                // We leave the old data intact -- it is not up-to-date, but at least it is correct.
+            }
+        } else {
+            // We leave the default `data_to_propose` -- we don't have the necessary input to compute a new one.
+        }
+    }
+
+    pub(crate) async fn get_best_header(&self) -> B::Header {
+        self.select_chain.best_chain().await.expect("No best chain")
+    }
+
+    pub(crate) async fn run(mut self, mut exit: oneshot::Receiver<()>) {
+        // We would like `best_block_in_session` to correspond to the highest ancestor of the `best_block` (this is what
+        // `select_chain` provides us with) up to the maximal height of `max_block_num`. This task periodically
+        // queries `select_chain` for the `best_block` and updates `best_block_in_session` accordingly. One optimization that it
```

Comment:
I think a lot of the optimizations below are again premature. Especially in the setting where we limit block production, but even in general, always `reduce`ing the best block to what we want should be fine (as most of the time this will be a noop). 

Not very sure about this one, I suspect in the current production regime it might be a problem if we fall behind too much, but even then I would rather reduce `REFRESH_INTERVAL`.

Reply:
I agree that if we stop block production than this optimization is not necessary, but so far we have not done it, so we don't our node to freeze simply because finalization stalled.

Reducing `REFRESH_INTERVAL` I think is not a great idea, because on average there is a contribution of `REFRESH_INTERVAL/2` in the overall latency of finalization. (that much time passes between the moment when a new block is imported and the moment our `DataProvider` is up-to-date).

Generally I hate this piece of code and the `SelectChain` thing in particular. But I don't feel I should make major changes here as part of this PR :/

Reply:
Also, I just want to mention that the optimization was always there :P I didn't touch it. I only refactored a little bit.

Reply:
Yeah, you are right, any changes here would be out of scope.